File: frontend/src/api/logs.js
/**
 * Log Analysis API — wraps all Phase 2 REST endpoints.
 * Matches the 10 routes registered in backend/routes/logs.py
 */
import api from './client'

// ── Log Upload & Management ─────────────────────────────────────────────────

export const logsAPI = {
  /**
   * Upload a raw log file for analysis.
   * POST /api/v1/logs/upload
   * @param {File} file  - The log file object
   * @param {string} [logType] - Optional forced log type (e.g. 'linux_syslog')
   * @param {string} [description] - Optional human description
   */
  upload: (file, logType = null, description = '') => {
    const form = new FormData()
    form.append('file', file)
    if (logType) form.append('force_log_type', logType)   // backend param name
    if (description) form.append('description', description)
    return api.post('/api/v1/logs/upload', form, {
      headers: { 'Content-Type': 'multipart/form-data' },
      timeout: 120_000, // 2 min — large files
    })
  },

  /**
   * Trigger the NLP pipeline on an already-uploaded log.
   * POST /api/v1/logs/:id/parse
   */
  parse: (id, forceLogType = null) =>
    api.post(`/api/v1/logs/${id}/parse`, forceLogType ? { force_log_type: forceLogType } : {}),

  /**
   * List all logs with optional filters.
   * GET /api/v1/logs
   */
  list: (params = {}) => api.get('/api/v1/logs', { params }),

  /**
   * Get single log metadata + pipeline summary.
   * GET /api/v1/logs/:id
   */
  get: (id) => api.get(`/api/v1/logs/${id}`),

  /**
   * Delete a log and all its derived data.
   * DELETE /api/v1/logs/:id
   */
  delete: (id) => api.delete(`/api/v1/logs/${id}`),


  /**
   * Get all parsed events for a log.
   * GET /api/v1/logs/:id/events
   */
  getEvents: (id, params = {}) => api.get(`/api/v1/logs/${id}/events`, { params }),

  /**
   * Get all IOCs extracted from a log.
   * GET /api/v1/logs/:id/iocs
   */
  getIOCs: (id, params = {}) => api.get(`/api/v1/logs/${id}/iocs`, { params }),
}

// ── IOC Explorer ────────────────────────────────────────────────────────────

export const iocsAPI = {
  /**
   * Search IOCs across all logs.
   * GET /api/v1/iocs/search
   */
  search: (params = {}) => api.get('/api/v1/iocs/search', { params }),

  /**
   * Get a single IOC entity by ID.
   * GET /api/v1/iocs/:id
   */
  get: (id) => api.get(`/api/v1/iocs/${id}`),
}

// ── Incidents ───────────────────────────────────────────────────────────────

export const incidentsAPI = {
  /**
   * List all incident events with pagination + filters.
   * GET /api/v1/incidents
   */
  list: (params = {}) => api.get('/api/v1/incidents', { params }),

  /**
   * Update incident status (open → investigating → resolved → false_positive).
   * PATCH /api/v1/incidents/:id/status
   */
  updateStatus: (id, status) =>
    api.patch(`/api/v1/incidents/${id}/status`, { status }),
}

// ── Statistics ──────────────────────────────────────────────────────────────

export const statisticsAPI = {
  /**
   * Get aggregated platform-wide statistics.
   * GET /api/v1/statistics
   */
  get: (params = {}) => api.get('/api/v1/statistics', { params }),
}

// ── Helpers ─────────────────────────────────────────────────────────────────

/** Severity badge colours — shared across all Phase 2 pages */
export const SEVERITY_CONFIG = {
  critical: { bg: 'bg-red-500/10', text: 'text-red-400', border: 'border-red-500/20', dot: 'bg-red-500' },
  high:     { bg: 'bg-orange-500/10', text: 'text-orange-400', border: 'border-orange-500/20', dot: 'bg-orange-500' },
  medium:   { bg: 'bg-amber-500/10', text: 'text-amber-400', border: 'border-amber-500/20', dot: 'bg-amber-500' },
  low:      { bg: 'bg-blue-500/10', text: 'text-blue-400', border: 'border-blue-500/20', dot: 'bg-blue-500' },
  info:     { bg: 'bg-slate-500/10', text: 'text-slate-400', border: 'border-slate-500/20', dot: 'bg-slate-500' },
}

/** Status badge colours for log processing status */
export const STATUS_CONFIG = {
  pending:    { text: 'text-slate-400', bg: 'bg-slate-500/10', border: 'border-slate-500/20' },
  processing: { text: 'text-cyan-400', bg: 'bg-cyan-500/10', border: 'border-cyan-500/20' },
  completed:  { text: 'text-emerald-400', bg: 'bg-emerald-500/10', border: 'border-emerald-500/20' },
  failed:     { text: 'text-red-400', bg: 'bg-red-500/10', border: 'border-red-500/20' },
}

/** Human-readable log type labels */
export const LOG_TYPE_LABELS = {
  linux_syslog:   'Linux Syslog',
  windows_event:  'Windows Event',
  apache_access:  'Apache Access',
  nginx_access:   'Nginx Access',
  sysmon:         'Sysmon (XML)',
  unknown:        'Unknown',
}

/** Format bytes → human readable string */
export function formatBytes(bytes) {
  if (!bytes || bytes === 0) return '0 B'
  const k = 1024
  const sizes = ['B', 'KB', 'MB', 'GB']
  const i = Math.floor(Math.log(bytes) / Math.log(k))
  return `${parseFloat((bytes / Math.pow(k, i)).toFixed(1))} ${sizes[i]}`
}

/** Format ISO date → short local string */
export function formatDate(iso) {
  if (!iso) return '—'
  return new Date(iso).toLocaleString(undefined, {
    month: 'short', day: 'numeric',
    hour: '2-digit', minute: '2-digit',
  })
}

File: backend/routes/logs.py
"""
SentinelX AI — Log & Incident Routes
========================================
All REST endpoints for Phase 2: Security Log Processing & NLP Engine.

Endpoint map:
  ── Logs ──────────────────────────────────────────────────────
  POST   /api/v1/logs/upload               Upload a raw log file
  POST   /api/v1/logs/{log_id}/parse       Trigger NLP pipeline
  GET    /api/v1/logs/                     Paginated log list
  GET    /api/v1/logs/{log_id}             Single log detail
  GET    /api/v1/logs/{log_id}/events      Parsed events for a log
  GET    /api/v1/logs/{log_id}/iocs        IOCs extracted from a log

  ── IOC Search ────────────────────────────────────────────────
  GET    /api/v1/iocs/search               Global IOC value search

  ── Incidents ─────────────────────────────────────────────────
  GET    /api/v1/incidents/                Paginated incident list
  PATCH  /api/v1/incidents/{id}/status     Update incident status

  ── Statistics ────────────────────────────────────────────────
  GET    /api/v1/statistics                Dashboard aggregate stats

Auth:
  - Upload / Parse: analyst or admin only
  - Read endpoints: any authenticated user
  - Status update: analyst or admin only
"""
from __future__ import annotations

import logging

from fastapi import APIRouter, Query, UploadFile, File, Form, status
from fastapi.responses import JSONResponse

from backend.dependencies import (
    AnalystOrAdmin,
    CurrentUser,
    DBSession,
)
from backend.schemas.common import APIResponse
from backend.schemas.logs import (
    IOCEntityListResponse,
    IncidentEventListResponse,
    LogParseRequest,
    LogParseResponse,
    LogUploadResponse,
    ParsedEventListResponse,
    SecurityLogListResponse,
    SecurityLogResponse,
    StatisticsResponse,
)
from backend.services.log_service import LogService

logger = logging.getLogger(__name__)

# ── Routers ────────────────────────────────────────────────────────────────────
logs_router      = APIRouter()
ioc_router       = APIRouter()
incident_router  = APIRouter()
stats_router     = APIRouter()


# =============================================================================
# LOGS
# =============================================================================

@logs_router.post(
    "/upload",
    summary="Upload a security log file",
    status_code=status.HTTP_201_CREATED,
    response_model=APIResponse[LogUploadResponse],
    dependencies=[AnalystOrAdmin],
)
async def upload_log(
    db:             DBSession,
    current_user:   CurrentUser,
    file:           UploadFile = File(..., description="Raw log file (txt, log, xml, evtx, csv)"),
    force_log_type: str | None = Form(
        default=None,
        description="Override auto-detection: linux_syslog | windows_event | apache_access | nginx_access | sysmon",
    ),
) -> JSONResponse:
    """
    Upload a raw security log file.

    - Stores the raw content and returns a `log_id`
    - Does **not** run analysis — call `POST /logs/{log_id}/parse` separately
    - File size limit: 100 MB
    - Supported formats: `.log`, `.txt`, `.xml`, `.evtx`, `.csv`
    """
    raw_bytes = await file.read()
    svc = LogService(db)
    result = await svc.upload_log(
        filename=file.filename or "upload.log",
        raw_content=raw_bytes,
        uploaded_by=getattr(current_user, "id", None),
        force_log_type=force_log_type,
    )
    return JSONResponse(
        status_code=status.HTTP_201_CREATED,
        content={
            "success": True,
            "message": result.message,
            "data":    result.model_dump(mode="json"),
        },
    )


@logs_router.post(
    "/{log_id}/parse",
    summary="Run NLP pipeline on an uploaded log",
    response_model=APIResponse[LogParseResponse],
    dependencies=[AnalystOrAdmin],
)
async def parse_log(
    log_id:  int,
    db:      DBSession,
    payload: LogParseRequest = LogParseRequest(),
) -> JSONResponse:
    """
    Trigger the full NLP processing pipeline on a stored log.

    Pipeline stages:
    1. **Detect** — identify log format automatically
    2. **Parse** — extract structured `NormalizedEvent` objects
    3. **Extract IOCs** — pull IPs, domains, hashes, URLs, usernames
    4. **Classify** — map to MITRE ATT&CK tactics and techniques
    5. **Persist** — store all results atomically in PostgreSQL

    Returns event counts, IOC counts, and processing time.
    """
    svc = LogService(db)
    result = await svc.process_log(
        log_id=log_id,
        force_log_type=payload.force_log_type,
    )
    return JSONResponse(
        content={
            "success": True,
            "message": result.message,
            "data":    result.model_dump(mode="json"),
        }
    )


@logs_router.get(
    "/",
    summary="List all security logs",
    response_model=APIResponse[SecurityLogListResponse],
)
async def list_logs(
    db:          DBSession,
    _:           CurrentUser,
    page:        int        = Query(default=1,    ge=1,           description="Page number"),
    page_size:   int        = Query(default=20,   ge=1,  le=100,  description="Items per page"),
    status:      str | None = Query(default=None, description="Filter by status: pending | processing | completed | failed"),
    log_type:    str | None = Query(default=None, description="Filter by log type"),
    uploaded_by: int | None = Query(default=None, description="Filter by uploader user ID"),
) -> JSONResponse:
    """
    Return a paginated list of security logs.

    Optionally filter by `status`, `log_type`, or `uploaded_by`.
    Results are ordered newest-first.
    """
    svc = LogService(db)
    result = await svc.list_logs(
        page=page,
        page_size=page_size,
        status=status,
        log_type=log_type,
        uploaded_by=uploaded_by,
    )
    return JSONResponse(
        content={
            "success": True,
            "message": f"Found {result.total} log(s).",
            "data":    result.model_dump(mode="json"),
        }
    )


@logs_router.get(
    "/{log_id}",
    summary="Get a single log record",
    response_model=APIResponse[SecurityLogResponse],
)
async def get_log(
    log_id: int,
    db:     DBSession,
    _:      CurrentUser,
) -> JSONResponse:
    """Fetch the details of a specific security log by ID."""
    svc = LogService(db)
    result = await svc.get_log(log_id)
    return JSONResponse(
        content={
            "success": True,
            "message": "OK",
            "data":    result.model_dump(mode="json"),
        }
    )


@logs_router.get(
    "/{log_id}/events",
    summary="Get parsed events for a log",
    response_model=APIResponse[ParsedEventListResponse],
)
async def get_log_events(
    log_id:             int,
    db:                 DBSession,
    _:                  CurrentUser,
    page:               int        = Query(default=1,    ge=1),
    page_size:          int        = Query(default=50,   ge=1, le=200),
    severity:           str | None = Query(default=None, description="critical | high | medium | low | info"),
    event_type:         str | None = Query(default=None, description="Filter by event type (e.g. 'Failed Login')"),
    source_ip:          str | None = Query(default=None, description="Filter by source IP address"),
    mitre_technique_id: str | None = Query(default=None, description="Filter by MITRE technique (e.g. T1110)"),
) -> JSONResponse:
    """
    Return structured events extracted from a processed log.

    Supports filtering by severity, event type, source IP, and MITRE technique.
    """
    svc = LogService(db)
    result = await svc.get_log_events(
        log_id,
        page=page,
        page_size=page_size,
        severity=severity,
        event_type=event_type,
        source_ip=source_ip,
        mitre_technique_id=mitre_technique_id,
    )
    return JSONResponse(
        content={
            "success": True,
            "message": f"Found {result.total} event(s).",
            "data":    result.model_dump(mode="json"),
        }
    )


@logs_router.get(
    "/{log_id}/iocs",
    summary="Get IOCs extracted from a log",
    response_model=APIResponse[IOCEntityListResponse],
)
async def get_log_iocs(
    log_id:   int,
    db:       DBSession,
    _:        CurrentUser,
    ioc_type: str | None = Query(default=None, description="Filter by type: ipv4 | domain | url | md5 | sha256 | ..."),
    page:     int        = Query(default=1,    ge=1),
    page_size:int        = Query(default=50,   ge=1, le=200),
) -> JSONResponse:
    """
    Return Indicators of Compromise (IOCs) extracted from a processed log.

    IOC types include: `ipv4`, `ipv6`, `domain`, `url`, `email`, `md5`,
    `sha1`, `sha256`, `cve`, `hostname`, `username`, `port`,
    `process_name`, `command_line`, `filename`, `file_path`.
    """
    svc = LogService(db)
    result = await svc.get_log_iocs(
        log_id,
        ioc_type=ioc_type,
        page=page,
        page_size=page_size,
    )
    return JSONResponse(
        content={
            "success": True,
            "message": f"Found {result.total} IOC(s).",
            "data":    result.model_dump(mode="json"),
        }
    )


# =============================================================================
# IOC SEARCH
# =============================================================================

@ioc_router.get(
    "/search",
    summary="Search IOCs by value across all logs",
    response_model=APIResponse[IOCEntityListResponse],
)
async def search_iocs(
    db:       DBSession,
    _:        CurrentUser,
    q:        str | None = Query(default=None, min_length=2, description="Value substring to search (e.g. '185.24', 'mimikatz'). Omit to list all IOCs."),
    ioc_type: str | None = Query(default=None,      description="Restrict search to a specific IOC type"),
    page:     int        = Query(default=1,  ge=1),
    page_size:int        = Query(default=50, ge=1, le=200),
) -> JSONResponse:
    """
    **Threat hunting endpoint** — search for an IOC value across all logs.

    - Omit `q` to list all IOCs (paginated, ordered by last_seen desc).
    - `?q=185.24.18` — find all events from a suspicious IP range
    - `?q=mimikatz` — find all command-line or filename IOCs mentioning mimikatz
    - `?q=T1059&ioc_type=command_line` — filter by type
    """
    svc = LogService(db)
    result = await svc.search_iocs(
        value=q,
        ioc_type=ioc_type,
        page=page,
        page_size=page_size,
    )
    msg = f"Found {result.total} IOC match(es) for '{q}'." if q else f"{result.total} IOC(s) total."
    return JSONResponse(
        content={
            "success": True,
            "message": msg,
            "data":    result.model_dump(mode="json"),
        }
    )


# =============================================================================
# INCIDENTS
# =============================================================================

@incident_router.get(
    "/",
    summary="List security incidents",
    response_model=APIResponse[IncidentEventListResponse],
)
async def list_incidents(
    db:        DBSession,
    _:         CurrentUser,
    page:      int        = Query(default=1,    ge=1),
    page_size: int        = Query(default=20,   ge=1, le=100),
    status:    str | None = Query(default=None, description="open | investigating | contained | closed | false_positive"),
    severity:  str | None = Query(default=None, description="critical | high | medium | low"),
) -> JSONResponse:
    """
    Return a paginated list of security incidents.

    Incidents are auto-created when the NLP pipeline detects HIGH+ severity events.
    Analysts can also create and manage incidents manually.
    """
    svc = LogService(db)
    result = await svc.get_incidents(
        page=page,
        page_size=page_size,
        status=status,
        severity=severity,
    )
    return JSONResponse(
        content={
            "success": True,
            "message": f"Found {result.total} incident(s).",
            "data":    result.model_dump(mode="json"),
        }
    )


@incident_router.patch(
    "/{incident_id}/status",
    summary="Update incident status",
    dependencies=[AnalystOrAdmin],
    response_model=APIResponse[dict],
)
async def update_incident_status(
    incident_id: int,
    db:          DBSession,
    status:      str = Query(..., description="New status: open | investigating | contained | closed | false_positive"),
) -> JSONResponse:
    """
    Update the investigation status of an incident.

    Valid transitions:
    `open` → `investigating` → `contained` → `closed`
    Any status may be set to `false_positive` to dismiss it.
    Closing an incident automatically stamps `closed_at`.
    """
    svc = LogService(db)
    result = await svc.update_incident_status(incident_id, status=status)
    return JSONResponse(
        content={
            "success": True,
            "message": f"Incident {incident_id} status updated to '{status}'.",
            "data":    result.model_dump(mode="json"),
        }
    )


# =============================================================================
# STATISTICS
# =============================================================================

@stats_router.get(
    "/",
    summary="Get platform-wide statistics",
    response_model=APIResponse[StatisticsResponse],
)
async def get_statistics(
    db: DBSession,
    _:  CurrentUser,
) -> JSONResponse:
    """
    Return aggregate statistics for the SentinelX dashboard.

    Includes:
    - Log counts by status (pending / processing / completed / failed)
    - Event counts with severity breakdown (critical → info)
    - IOC counts and type distribution
    - Incident counts by status and severity
    - Top attacker IPs
    - Top MITRE ATT&CK techniques
    """
    svc = LogService(db)
    result = await svc.get_statistics()
    return JSONResponse(
        content={
            "success": True,
            "message": "OK",
            "data":    result.model_dump(mode="json"),
        }
    )
